v2/traversal/resumption.go

```diff
@@ -0,0 +1,257 @@
+
+import (
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/ipld/go-car/v2/internal/loader"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking"
+	"github.com/ipld/go-ipld-prime/traversal"
+)
+
+type pathNode struct {
+	link     datamodel.Link
+	offset   uint64
+	children map[datamodel.PathSegment]*pathNode
+}
+
+func newPath(link datamodel.Link, at uint64) *pathNode {
+	return &pathNode{
+		link:     link,
+		offset:   at,
+		children: make(map[datamodel.PathSegment]*pathNode),
+	}
+}
+
+func (pn pathNode) addPath(p []datamodel.PathSegment, link datamodel.Link, at uint64) {
+	if len(p) == 0 {
+		return
+	}
+	if _, ok := pn.children[p[0]]; !ok {
+		child := newPath(link, at)
+		pn.children[p[0]] = child
+	}
+	pn.children[p[0]].addPath(p[1:], link, at)
+}
+
+func (pn pathNode) allLinks() []datamodel.Link {
+	if len(pn.children) == 0 {
+		return []datamodel.Link{pn.link}
+	}
+	links := make([]datamodel.Link, 0)
+	if pn.link != nil {
+		links = append(links, pn.link)
+	}
+	for _, v := range pn.children {
+		links = append(links, v.allLinks()...)
+	}
+	return links
+}
+
+// getPaths returns reconstructed paths in the tree rooted at 'root'
+func (pn pathNode) getLinks(root datamodel.Path) []datamodel.Link {
+	segs := root.Segments()
+	switch len(segs) {
+	case 0:
+		if pn.link != nil {
+			return []datamodel.Link{pn.link}
+		}
+		return []datamodel.Link{}
+	case 1:
+		// base case 1: get all paths below this child.
+		next := segs[0]
+		if child, ok := pn.children[next]; ok {
+			return child.allLinks()
+		}
+		return []datamodel.Link{}
+	default:
+	}
+
+	next := segs[0]
+	if _, ok := pn.children[next]; !ok {
+		// base case 2: not registered sub-path.
+		return []datamodel.Link{}
+	}
+	return pn.children[next].getLinks(datamodel.NewPathNocopy(segs[1:]))
+}
+
+var errInvalid = fmt.Errorf("invalid path")
+
+func (pn pathNode) offsetAfter(root datamodel.Path) (uint64, error) {
+	// we look for offset of next sibling.
+	// if no next sibling recurse up the path segments until we find a next sibling.
+	segs := root.Segments()
+	if len(segs) == 0 {
+		return 0, errInvalid
+	}
+	// see if this path is a child.
+	chld, ok := pn.children[segs[0]]
+	if !ok {
+		return 0, errInvalid
+	}
+	closest := chld.offset
+	// try recursive path
+	if len(segs) > 1 {
+		co, err := chld.offsetAfter(datamodel.NewPathNocopy(segs[1:]))
+		if err == nil {
+			return co, err
+		}
+	}
+	// find our next sibling
+	var next uint64 = math.MaxUint64
+	var nc *pathNode
+	for _, v := range pn.children {
+		if v.offset > closest && v.offset < next {
+			next = v.offset
+			nc = v
+		}
+	}
+	if nc != nil {
+		return nc.offset, nil
+	}
+
+	return 0, errInvalid
+}
+
+// TraverseResumer allows resuming a progress from a previously encountered path in the selector.
+type TraverseResumer interface {
+	RewindToPath(from datamodel.Path) error
```

Comment:
am I right in assuming we don't have a use-case for this at the moment, that we're just supporting it because it's not a difficult extension of doing it to an offset?

got thoughts on what a use-case for might me?

Reply:
This is probably the most useful thing to have pointed out. I think the more useful method to expose here will be a `RewindToOffset` in addition to the current `RewindToPath`. I'll work on adding that